Core-1249  Elasticsearch Drain Utility

athompson-r7 opened the pull request:
The purpose of this utility is to drain documents from Elasticsearch nodes in an AutoScaling Group.

This will help you do Elasticsearch node replacement while keeping the cluster healthy.

rvanburen-r7 commented:
Looks good to me, but you may want a second pair of eyes.

fmitchell-r7 commented on Rakefile:
```diff
@@ -1,2 +1,68 @@
+
+ENV['TEST_CLUSTER_NODES'] = '1'
+ENV['TEST_CLUSTER_COMMAND'] = elasticsearch_command
+
```
Maybe expose ES version as an environment variable too?

fmitchell-r7 commented on README.md:
```diff
@@ -1,30 +1,30 @@
+This will help you do Elasticsearch node replacement while keeping the cluster healthy.
 
-Add this line to your application's Gemfile:
+Consider the following deployment procedure:
```
Might want to include a note that this is a useful way to do an upgrade if you can't update binaries in place, like when you bake versioned AMIs.

athompson-r7 replied:
Addressed in 49ac49c.

fmitchell-r7 commented on lib/elasticsearch/drain.rb:
```diff
@@ -1,7 +1,113 @@
+
+    # EC2 AutoScaling Client
+    def asg
+      AutoScaling.new(@asg_name, @region)
```
Worth memoizing like `client` does?

~~~ruby
@asg ||= AutoScaling.new(@asg_name, @region)
~~~

fmitchell-r7 commented on lib/elasticsearch/drain.rb:
```diff
@@ -1,7 +1,113 @@
+    #
+    # @return [Elasticsearch::API::Cluster] Elasticsearch cluster client
+    def cluster
+      Cluster.new(client)
```
Worth memoizing?
~~~ruby
@cluster_client ||= Cluster.new(client)
~~~

fmitchell-r7 commented on lib/elasticsearch/drain.rb:
```diff
@@ -1,7 +1,113 @@
+    def drain
+      fail 'Cluster is unhealthy' unless cluster.healthy?
+      instances = asg.instances
+      nodes_to_drain = nodes.nodes_in_asg(reload: true, instances: instances)
```
Can you call `self.active_nodes_in_asg` here instead?

athompson-r7 commented:
@fmitchell-r7 I'm still looking at one last issue with memoizing the asg client; in the mean time can you take a look at the recent changes.  I refactored the cli interface, like I should have in the first place.  I moved things around so that I could use the thor magic. 

https://github.com/rapid7/elasticsearch-drain/compare/26198728790ed6425ce5028051506728cc43c0f5...CORE-1249

fmitchell-r7 commented:
:+1:
